---
title: Quantifying Gene Expression
---

In order to draw conclusions about gene expression from our reads, we need to first identify which region of the genome reads originated from and then quantify the number of reads from that region. There are many tools available to carry out quantification, and we will not cover the arguments for and against each tool. We encourage you to carry out a literature search, and discuss with experienced individuals in your field, before committing to a particular quantification procedure. Here we will give a brief overview of two approaches with considerably different underlying methodology, give an example tool for each, and then demonstrate one tool for the purpose of progressing to the next stage of our workflow. 

The first approach to read quantification involves aligning reads to the genome and then counting the number of reads that are aligned to certain genomic features (*e.g.,* exons). 

The second approach we will introduce involves creating 'pseudocounts'. In this method, reads are *not* aligned to the genome. Instead, they are pseudoaligned to kmers within the transcriptome and their abundance is estimated. 
**This explanation needs work, researching Salmon's methods currently**

In today's workflow we will demonstrate genome alignment and counting, and will provide examples of other workflows that use pseudocounting and abundance estimation. 

**Can copy over our existing workshop material into this section without needing to modify much**


